### PR TITLE
Improve backend routes auth

### PR DESCRIPTION
**Pull request by alesshh**
# Problem
Our backend routes are unsafe and can be accessible without login. 😱 

## Solution
This PR will require user authentication on all private routes.

## Expected behavior

- The route `/` redirect users to rooms screen if already logged in
- The routes `/morpheus*` redirect unlogged users to login screen
- The route `/new` redirect unlogged users to login screen
- The route `/remove` redirect unlogged users to login screen
- The route `/rooms` send a 401 if the user is not logged in

**The socket layer remains unprotected because of we pass the user data to the socket, without verifications.**
The middleware used to store cookie sessions don't work with socket.io middlewares. 
We need to create a cookie session reader/writer, able to works with socket.io and express middlewares. I will open an issue to solve it in the future.

### CODE REVIEW

File: backend/app/app.routes.js [hunk 1]
```diff
@@ -1,25 +1,27 @@
...
-  res.render("index", {
-    isAuthenticated: !!req.session.currentUser,
-    error: req.query.error,
-  });
+  if (auth.isUserLoggedIn(req)) {
+    return res.redirect("/morpheus");
+  }
+
+  return res.render("index", { error: req.query.error });
 });
```
Comment (dvinciguerra) [inline backend/app/app.routes.js]: Isn't it better to add a `home.controller.js` to handle this logic?

Something like this:

```javascript
import Home from './controllers/home.controller';

router.get("/", Home.index);
```

File: backend/app/app.auth.js [hunk 2]
```diff
@@ -9,7 +9,7 @@ passport.use(
     {
       clientID: config.GOOGLE_CLIENT_ID,
       clientSecret: config.GOOGLE_SECRET,
-      callbackURL: config.GOOGLE_CALLBACK_URL,
+      callbackURL: config.GOOGLE_CALLBACK_URL
     },
```
Comment (dvinciguerra) [inline backend/app/app.auth.js]: What do you think about this approach?

```javascript
import {
  GOOGLE_CLIENT_ID as clientID,
  GOOGLE_SECRET as clientSecret,
  GOOGLE_CALLBACK_URL as callbackURL
} from "./app.config";

...

passport.use(
  new GooglePassport.Strategy(
    { clientID, clientSecret, callbackURL },
    (accessToken, refreshToken, profile, cb) => {
		...
  })
}
```

File: backend/app/app.auth.js [hunk 3]
```diff
@@ -32,8 +32,53 @@ passport.deserializeUser((user, done) => {
 });
```
Comment (dvinciguerra) [inline backend/app/app.auth.js]: What do you think about transform this object into a "module class" using static methods?

```javascript
class Auth {
  static initialize() { ... }

  static session() { ... }

  static isUserLoggedIn(req) { ... }

  static currentUser(req) { ... }

  static authenticate(opts) { ... }

  static login(req, user) { ... }

  static logout(req) { ... }
}

module.exports = Auth;
```

Look more organized in my opinion and don't change the usage (tests and imports).

File: backend/app/app.routes.js (hunk 1 above)
Comment (alesshh) [reply backend/app/app.routes.js]: I think we can extract all route functions  on a separated pr. What do you think?

Comment (alesshh) [reply backend/app/app.routes.js]: The index is the login page 😅 .
I don't like javascript classes for imports as much. My personal preference is to export only functions and explicit import what do you want.

What do you think to discuss about the backend structure / code on a issue?

File: backend/app/app.auth.js (hunk 2 above)
Comment (alesshh) [reply backend/app/app.auth.js]: Better! 🚀

File: backend/app/app.auth.js (hunk 3 above)
Comment (alesshh) [reply backend/app/app.auth.js]: I don't like javascript classes for imports as much. My personal preference is to export only functions and explicit import what do you want.

What do you think to discuss about the backend structure / code on a issue?

File: backend/app/app.routes.js (hunk 1 above)
Comment (dvinciguerra) [reply backend/app/app.routes.js]: Perfect, I agree!!! 👏

File: backend/app/app.auth.js (hunk 3 above)
Comment (dvinciguerra) [reply backend/app/app.auth.js]: Sure man! =)

Comment (megatroom) [reply backend/app/app.auth.js]: I also prefer not to use classes in JavaScript, but either way doesn't make sense to me in having a class with only static methods, it run away from OOP.

File: backend/app/app.auth.js [hunk 4]
```diff
@@ -32,8 +32,53 @@ passport.deserializeUser((user, done) => {
...
+
+  isUserLoggedIn(req) {
+    return !!this.currentUser(req);
+  },
+
+  currentUser(req) {
+    return req.user;
+  },
+
+  authenticate(opts) {
```
Comment (megatroom) [inline backend/app/app.auth.js]: What do you think about moving this method to a file inside a directory of middlewares? Example:

`backend/app/middlewares/authenticate.js`

Is not clear to me that this method is a middleware. I only found out when I saw the usage on the route.

What do you think?

File: backend/app/app.auth.js [hunk 5]
```diff
@@ -32,8 +32,53 @@ passport.deserializeUser((user, done) => {
...
+    const { loginURL = null } = opts || {};
```
Comment (megatroom) [inline backend/app/app.auth.js]:
```js
const loginURL = opts && opts.loginURL;
```

What do you think of doing this way?

File: frontend/src/pages/login/login.js [hunk 6]
```diff
@@ -14,85 +14,44 @@ import "bootstrap/dist/css/bootstrap.css";
...
-                    }}
-                  />
-                  {error && (
-                    <p className={clsx("text-danger", styles.error)}>
-                      {error}
-                    </p>
-                  )}
-                </div>
-                <Footer />
+export function Login() {
```
Comment (megatroom) [inline frontend/src/pages/login/login.js]: :heart:

File: frontend/src/pages/login/login.js [hunk 7]
```diff
@@ -14,85 +14,44 @@ import "bootstrap/dist/css/bootstrap.css";
...
+  const [error, setError] = useState(null);
```
Comment (megatroom) [inline frontend/src/pages/login/login.js]: :heart:

**Comment by alesshh** [comment]
Hey guys, given your feedbacks I made some changes:
-  Extracted auth and passport responsibility to a service `app/services/auth` that encapsulate login and session passport logic, this way we can easily add new auth strategies with a single config.
- Refactored config file to functions,  this way, we do not need to use ugly constants on app setup, we can group configs by context;
- Added morgan logger to better development;
- Created an object to map route paths on route files;

File: backend/app/app.server.js [hunk 8]
```diff
@@ -2,66 +2,61 @@ import express from "express";
...
+  getCustomStyle,
+  getSessionConfig,
+  getEnvironment,
+  shouldEnforceSSL
 } from "./app.config";
 
 const app = express();
 
-app.locals.CUSTOM_STYLE = CUSTOM_STYLE;
+app.locals.CUSTOM_STYLE = getCustomStyle();
```
Comment (angeliski) [inline backend/app/app.server.js]: I think the variable CUSTOM_STYLE is no longer necessary.
https://github.com/ResultadosDigitais/matrix/pull/167/files
Or maybe is missing feature in Morpheus layout

Comment (alesshh) [reply backend/app/app.server.js]: 🔪

Comment (juliemar) [reply backend/app/app.server.js]: Yes... we can remove this variable :) my fault :disappointed:

**Pull request merged**